app/controllers/games_controller.rb

```diff
@@ -1,74 +1,48 @@
 class GamesController < ApplicationController
-  before_action :set_game, only: [:show, :edit, :update, :destroy]
+  before_action :set_game, only: [:add_player, :deal, :player_action, :show]
 
   def add_player
-    unless @_current_user.nil?
-      set_game
-      state = @game.load_state
-      if state[:players].size < 5
-        unless state[:players].include?(@_current_user)
-          if state[:cards_in_play]
-            redirect_to @game, notice: 'Cant join once the game has already started'
-            return
-          end
-          state[:players].push @_current_user
-          while state[:names].include? params[:username]
-            # make sure username is unique by appending random numbers
-            params[:username] += rand(10).to_s
-          end
-          state[:names].push params[:username]
-          @game.save_state state
-          if @game.save
-            redirect_to @game, notice: 'Joined the game!'
-          else
-            redirect_to @game, notice: 'Failed to join the game'
-          end
-        else
-          redirect_to @game, notice: 'Youre already in the game'
-        end
-      else
-        redirect_to @game, notice: 'There are too many players in the game already'
-      end
+    if @game.state_data[:players].size > 5
+      redirect_to @game, notice: 'There are too many players in the game already'
+    elsif @game.has_player?(current_user)
+      redirect_to @game, notice: 'Youre already in the game'
+    elsif !@game.waiting_for_players?
+      redirect_to @game, notice: 'Cant join once the game has already started'
+    else
+      session[:player_name] = @game.add_player(current_user, params[:username])
+      redirect_to @game, notice: "Welcome to the game"
     end
   end
-  
+
   def deal
-    set_game
-    state = @game.load_state
-    if state[:players].size < 3 or state[:players].size > 5
-      redirect_to @game, notice: 'Must have between 3 and 5 players to start'
-      return
-    end
-    if state[:players].first == @_current_user
-      if state[:cards_in_play].nil?
-        @game.save_state @game.deal_cards(state)
-        @game.save
-        redirect_to @game, notice: 'Game has started!'
-      else
-        redirect_to @game, notice: 'Game has already started'
-      end
-    else
-      redirect_to @game, notice: 'Only the creator of the game can start it'
-    end
+    message = if @game.state_data[:players].size < 3 or @game.state_data[:players].size > 5
```

Comment:
I prefer the format in the above method (add_player), calling redirect_to in each branch. The extra indentation here is off putting.